cockatrice/src/dlg_connect.cpp

```diff
@@ -156,26 +168,96 @@ void DlgConnect::actSaveConfig()
                                               passwordEdit->text(), savePasswordCheckBox->isChecked());
     }
 
-    rebuildComboBoxList();
+    preRebuildComboBoxList();
 }
 
-void DlgConnect::rebuildComboBoxList()
+void DlgConnect::downloadPublicServers()
 {
     previousHosts->clear();
+    previousHosts->addItem(placeHolderText);
 
+    QUrl url(QString(PUBLIC_SERVERS_JSON));
+    auto *nam = new QNetworkAccessManager(this);
+    QNetworkReply *reply = nam->get(QNetworkRequest(url));
+    connect(reply, SIGNAL(finished()), this, SLOT(actFinishParsingDownloadedData()));
+}
+
+void DlgConnect::actFinishParsingDownloadedData()
+{
+    auto *reply = dynamic_cast<QNetworkReply *>(sender());
+    QNetworkReply::NetworkError errorCode = reply->error();
+
+    if (errorCode == QNetworkReply::NoError) {
+        // Get current saved hosts
+        UserConnection_Information uci;
+        savedHostList = uci.getServerInfo();
+
+        // List of public servers
+        // Will be populated below
+        QStringList publicServersAvailable;
+
+        // Downloaded data from GitHub
+        QString jsonData = QString(reply->readAll());
+        auto jsonMap = QtJson::Json::parse(jsonData).toMap();
+
+        // Servers available
+        auto publicServersJSONList = jsonMap["servers"].toList();
+        for (const auto &server : publicServersJSONList) {
+            // Data inside one server at a time
+            // server: [{ ... }, ..., { ... }]
+            const auto serverMap = server.toMap();
+
+            QString serverName = serverMap["name"].toString();
+            QString serverAddress = serverMap["host"].toString();
+            QString serverPort = serverMap["port"].toString();
+
+            publicServersAvailable.append(serverName);
+
+            // If this server is not found in the old list, add it
+            // Otherwise, do nothing (as the user already saved their
+            // information they want for this server!)
+            if (!savedHostList.contains(serverName)) {
+                settingsCache->servers().addNewServer(serverName, serverAddress, serverPort, "", "", false);
+            }
+        }
+
+        // If a server was removed from the public list,
+        // we will delete it from the local system
+        // as that server is no longer around.
+        for (auto server : savedHostList) {
+            QString serverName = server.getSaveName();
+            if (publicServersAvailable.indexOf(serverName) < 0) {
+                settingsCache->servers().removeServer(serverName);
+            }
+        }
```

Comment:
I see a couple problematic points with this approach:
 * if a server gets updated/edited, the change won't be picked up
 * if the user added a private/custom server (eg. i usually connect to 127.0.0.1 for testing), the custom server will be removed.

The first point could be resolved just updating the server address/port if the name matches.
The second point is a bit more difficult: should we handle 2 different types of server, "official" servers and user-defined server?
We could just add an "official" field to each server and only update those?

Reply:
I think the problem with the "official" field would be that at those users who have a configuration generated by a previous version of the program we couldn't differentiate between an "official" and a "custom" server - we either have to delete both or neither.
But I think we could solve this problem if we were to add a tag to the json file storing whether a specific server is "dead" or not. If it is dead and the user's config contains it then it should be deleted, otherwise it should be added, or if it already exists updated. The only drawback would be that we mustn't remove obsolete servers from the json file in this case.